### PR TITLE
Add torchao kernels to llama runner

### PR DESCRIPTION
**Pull request by metascroy**
Setup ET by following directions here: https://pytorch.org/executorch/stable/getting-started-setup

After ET is setup, we can build the Llama runner with torchao kernels.

Step 1 (build ET)
```
cmake -DPYTHON_EXECUTABLE=python \
    -DCMAKE_INSTALL_PREFIX=cmake-out \
    -DEXECUTORCH_ENABLE_LOGGING=1 \
    -DCMAKE_BUILD_TYPE=Release \
    -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON \
    -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON \
    -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON \
    -DEXECUTORCH_BUILD_XNNPACK=ON \
    -DEXECUTORCH_BUILD_MPS=ON \
    -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
    -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON \
    -DEXECUTORCH_BUILD_KERNELS_CUSTOM=ON \
    -Bcmake-out .
cmake --build cmake-out -j16 --target install --config Release
```

Step2 (build runner with torchao):
```
cmake -DPYTHON_EXECUTABLE=python \
    -DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
    -DCMAKE_INSTALL_PREFIX=cmake-out \
    -DCMAKE_BUILD_TYPE=Release \
    -DEXECUTORCH_BUILD_KERNELS_CUSTOM=ON \
    -DEXECUTORCH_BUILD_TORCHAO=ON \
    -DTORCHAO_BUILD_EXECUTORCH_OPS=ON \
    -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON \
    -DEXECUTORCH_BUILD_XNNPACK=ON \
    -DEXECUTORCH_BUILD_MPS=ON \
    -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON \
    -Bcmake-out/examples/models/llama \
    examples/models/llama
cmake --build cmake-out/examples/models/llama -j16 --target install --config Release
```

Step3 (install runner requirements):
```
sh examples/models/llama/install_requirements.sh
```
    
Step3 (export model):
```
CMAKE_INSTALL_PREFIX=$PWD/cmake-out python -m examples.models.llama.export_llama --checkpoint /path/to/model.pth --params  /path/to/params.json -kv --use_sdpa_with_kv_cache -qmode "torchao:8da3w" --group_size 128 -E "torchao:3,32" --disable_dynamic_shape --metadata '{"get_bos_id":128000, "get_eos_ids":[128009, 128001]}' --output_name /path/to/output.pte
```
The above quantizes the embeddings to 3 bits with groupsize 32 (-E torchao:3,32) and quantizes the linear layers with 8-bit dynamically quantized activations and weights to 3 bits with groupsize 128 "-qmode torchao:8da3w".  You can play around with other quantization schemes.  Starting with 4-bit (instead of 3-bit) is a good starting place for model quality.  torchao supports 1-7 bit quantization for both linear and embedding layers.

Step4 (run model):
```
cmake-out/examples/models/llama/llama_main --model_path=/path/to/output.pte --tokenizer_path=/path/to/tokenizer.model --prompt="Once upon a time,"
```

Note: you can also export quantized model pte files with torchchat (https://github.com/pytorch/torchchat/blob/main/docs/quantization.md#executorch-1) and then run the PTE files using Step4.

### CODE REVIEW

File: examples/models/llama2/CMakeLists.txt [hunk 1]
```diff
@@ -121,6 +121,11 @@ if(EXECUTORCH_BUILD_KERNELS_CUSTOM)
   list(APPEND link_libraries custom_ops)
 endif()
 
+if(EXECUTORCH_BUILD_TORCHAO)
+  list(APPEND link_libraries "$<LINK_LIBRARY:WHOLE_ARCHIVE,${CMAKE_CURRENT_BINARY_DIR}/../../../lib/libtorchao_ops_executorch.a>")
```
Comment (metascroy) [inline examples/models/llama2/CMakeLists.txt:125]: Exporting targets with config in torchao makes this bit nicer.

Comment (larryliu0820) [reply examples/models/llama2/CMakeLists.txt:125]: You already called `add_subdirectory` so here you can do something like this:

```
target_link_options(torchao_ops_executorch INTERFACE -Wl,--whole-archive -Wl,--no-whole-archive)

```

Or actually this line should live in torchao/experimental/CMakeLists.txt.

File: examples/models/llama/CMakeLists.txt [hunk 2]
```diff
@@ -121,6 +123,13 @@ if(EXECUTORCH_BUILD_KERNELS_CUSTOM)
   list(APPEND link_libraries custom_ops)
 endif()
 
+if(EXECUTORCH_BUILD_TORCHAO)
+  set(TORCHAO_BUILD_EXECUTORCH_OPS ON)
+  add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/ao/torchao/experimental ${CMAKE_CURRENT_BINARY_DIR}/../../../third-party/ao/torchao/experimental)
```
Comment (larryliu0820) [inline examples/models/llama/CMakeLists.txt:128]: I think you can do `${EXECUTORCH_ROOT}/third-party/ao/torchao/experimental`

File: examples/models/llama/source_transformation/quantize.py [hunk 3]
```diff
@@ -70,6 +72,69 @@ def quantize(  # noqa C901
...
         return WeightOnlyInt8QuantHandler(model).quantized_model()
+    elif qmode.startswith("torchao:"):
+        import glob
+        import os
+
+        libs = glob.glob(
+            os.path.abspath(
+                os.path.join(
+                    os.path.dirname(__file__),
+                    "../../../../cmake-out/third-party/ao/torchao/experimental/libtorchao_ops_aten.*",
```
Comment (larryliu0820) [inline examples/models/llama/source_transformation/quantize.py:83]: uh this hardcoded path is not ideal. Can we add install() in torchao/experimental/CMakeLists.txt and then we can find the installed library in CMAKE_INSTALL_PREFIX?

File: examples/models/llama/CMakeLists.txt (hunk 2 above)
Comment (metascroy) [reply examples/models/llama/CMakeLists.txt:128]: But one is the source directory and the other the binary directory?

File: examples/models/llama/source_transformation/quantize.py (hunk 3 above)
Comment (metascroy) [reply examples/models/llama/source_transformation/quantize.py:83]: If I understand, you want the path to be something like:

> "{CMAKE_INSTALL_PREFIX}/lib/libtorchao_ops_aten.*"

How is the python script going to know what CMAKE_INSTALL_PREFIX is?  Do you want to user to define this as an environment variable?

File: examples/models/llama/CMakeLists.txt (hunk 2 above)
Comment (larryliu0820) [reply examples/models/llama/CMakeLists.txt:128]: Yeah, I meant for the first one we can make it simpler. But it's up to you

**Review by larryliu0820 — APPROVED** [review]

**Review by kimishpatel — APPROVED** [review]

**Pull request merged**